File: juan_vivi/infraestructura/escritor.py
"""
Escritor — infraestructura de salida.
Genera el Excel de planilla masivos replicando el formato del archivo de referencia:
  - Fila 1: encabezados de grupo combinados (A1:D1, E1:H1, I1:T1)
  - Fila 2: encabezados de columnas individuales, altura 30pt, borde inferior thin
  - Filas 3+: datos
  - Anchos de columna según el archivo original
  - Fuente Calibri 11, sin negritas (replicando el original)
"""
import os
from datetime import datetime

import openpyxl
from openpyxl.styles import Font, Alignment, Border, Side, PatternFill
from openpyxl.utils import get_column_letter

from ..dominio.modelos import FilaSalida


# ── Anchos de columna (extraídos del archivo de referencia) ──────
ANCHOS_COLUMNA = {
    1:  13.0,       # A — RUT c1
    2:  41.14,      # B — Apellido Paterno c1
    3:  20.57,      # C — Apellido Materno c1
    4:  33.57,      # D — Nombres c1
    5:  13.0,       # E — RUT c2
    6:  41.14,      # F — Apellido Paterno c2
    7:  20.57,      # G — Apellido Materno c2
    8:  33.57,      # H — Nombres c2
    9:  13.86,      # I — Tasación Fiscal
    10: 15.0,       # J — Precio de Venta
    11: 12.86,      # K — Patente
    12: 8.0,        # L — Tipo
    13: 10.0,       # M — Marca
    14: 14.0,       # N — Modelo
    15: 6.0,        # O — Año
    16: 8.0,        # P — Color
    17: 10.0,       # Q — Motor
    18: 12.0,       # R — Chasis
    19: 10.0,       # S — Serie
    20: 10.0,       # T — Vin
    21: 19.57,      # U — Código de Operación
}

# ── Encabezados ───────────────────────────────────────────────────
GRUPOS_F1 = [
    ("Compareciente 1 / Vendedor",  "A1", "D1"),
    ("Compareciente 2 / Comprador", "E1", "H1"),
    ("Vehículo",                    "I1", "T1"),
    # U1 va solo (sin combinar)
]

ENCABEZADOS_F2 = [
    "RUT",
    "Apellido Paterno /\nRazón Social",
    "Apellido Materno",
    "Nombres",
    "RUT",
    "Apellido Paterno /\nRazón Social",
    "Apellido Materno",
    "Nombres",
    "Tasación Fiscal",
    "Precio de Venta",
    "Patente",
    "Tipo",
    "Marca",
    "Modelo",
    "Año",
    "Color",
    "Motor",
    "Chasis",
    "Serie",
    "Vin",
    "Código de Operación",
]


# ── Helpers de estilo ─────────────────────────────────────────────

def _fuente(bold: bool = False) -> Font:
    return Font(name="Calibri", size=11, bold=bold)

def _borde_inferior() -> Border:
    delgado = Side(border_style="thin")
    return Border(bottom=delgado, top=delgado)

def _centrado() -> Alignment:
    return Alignment(horizontal="center", vertical="center", wrap_text=True)

def _izquierda() -> Alignment:
    return Alignment(horizontal="left", vertical="center")


# ── Función principal ─────────────────────────────────────────────

def escribir_planilla(filas_salida: list, ruta_destino: str) -> None:
    """
    Genera el Excel de planilla masivos en ruta_destino.
    Recibe lista[FilaSalida] ya transformadas.
    """
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Hoja1"

    # ── Anchos de columna ────────────────────────────────────────
    for num_col, ancho in ANCHOS_COLUMNA.items():
        ws.column_dimensions[get_column_letter(num_col)].width = ancho

    # ── Fila 1: encabezados de grupo combinados ───────────────────
    for texto, inicio, fin in GRUPOS_F1:
        ws.merge_cells(f"{inicio}:{fin}")
        celda = ws[inicio]
        celda.value     = texto
        celda.font      = _fuente()
        celda.alignment = _centrado()

    # U1 — sin combinar
    ws["U1"].value     = "Datos OT"
    ws["U1"].font      = _fuente()
    ws["U1"].alignment = _izquierda()

    ws.row_dimensions[1].height = 18

    # ── Fila 2: encabezados individuales ─────────────────────────
    for col_idx, titulo in enumerate(ENCABEZADOS_F2, start=1):
        celda = ws.cell(row=2, column=col_idx)
        celda.value     = titulo
        celda.font      = _fuente()
        celda.alignment = _centrado()
        celda.border    = _borde_inferior()

    ws.row_dimensions[2].height = 30

    # ── Filas de datos ────────────────────────────────────────────
    for fila_idx, fila in enumerate(filas_salida, start=3):
        valores = [
            # Compareciente 1
            fila.c1_rut,
            fila.c1_apellido_paterno,
            fila.c1_apellido_materno,
            fila.c1_nombres,
            # Compareciente 2
            fila.c2_rut,
            fila.c2_apellido_paterno,
            fila.c2_apellido_materno,
            fila.c2_nombres,
            # Vehículo
            fila.tasacion_fiscal,
            fila.precio_venta,
            fila.patente,
            fila.tipo_vehiculo,
            fila.marca,
            fila.modelo,
            fila.anio,
            fila.color,
            fila.motor,
            fila.chasis,
            fila.serie,
            fila.vin,
            # Datos OT
            fila.codigo_operacion,
        ]
        for col_idx, valor in enumerate(valores, start=1):
            celda = ws.cell(row=fila_idx, column=col_idx, value=valor)
            celda.font      = _fuente()
            celda.alignment = _izquierda()

    # ── Guardar ───────────────────────────────────────────────────
    os.makedirs(os.path.dirname(ruta_destino) or ".", exist_ok=True)
    wb.save(ruta_destino)


def nombre_archivo_salida(carpeta_destino: str = ".", prefijo_sp: str = "") -> str:
    """
    Genera el nombre del archivo de salida.
    Si se entrega prefijo_sp (ej. "3206"), el nombre será:
      Repertorios Masivo SP 3206.xlsx
    Si son varios archivos:
      Repertorios Masivo SP 3206_3217.xlsx
    Sin prefijo:
      planilla_masivos_2026-07-27_09-14.xlsx
    """
    if prefijo_sp:
        nombre = f"Repertorios Masivo SP {prefijo_sp}.xlsx"
    else:
        marca = datetime.now().strftime("%Y-%m-%d_%H-%M")
        nombre = f"planilla_masivos_{marca}.xlsx"
    return os.path.join(carpeta_destino, nombre)

File: src-tauri/src/escritor.rs
use chrono::Local;
use rust_xlsxwriter::{Format, FormatAlign, FormatBorder, Workbook};

use crate::modelos::FilaSalida;

// ─── Anchos de columna (índice 0-based, ancho) ───────────────────────────────
const ANCHOS_COLUMNA: &[(u16, f64)] = &[
    (0, 13.0),
    (1, 41.14),
    (2, 20.57),
    (3, 33.57),
    (4, 13.0),
    (5, 41.14),
    (6, 20.57),
    (7, 33.57),
    (8, 13.86),
    (9, 15.0),
    (10, 12.86),
    (11, 8.0),
    (12, 10.0),
    (13, 14.0),
    (14, 6.0),
    (15, 8.0),
    (16, 10.0),
    (17, 12.0),
    (18, 10.0),
    (19, 10.0),
    (20, 19.57),
];

// ─── Encabezados individuales (fila 2) ───────────────────────────────────────
const ENCABEZADOS: &[&str] = &[
    "RUT",
    "Apellido Paterno",
    "Apellido Materno",
    "Nombres",
    "RUT",
    "Apellido Paterno",
    "Apellido Materno",
    "Nombres",
    "Tasación Fiscal",
    "Precio Venta",
    "Patente",
    "Tipo Vehículo",
    "Marca",
    "Modelo",
    "Año",
    "Color",
    "N° Motor",
    "N° Chasis",
    "N° Serie",
    "VIN",
    "Código Operación",
];

// ─── Escribir planilla ────────────────────────────────────────────────────────

pub fn escribir_planilla(filas: &[FilaSalida], ruta: &str) -> anyhow::Result<()> {
    let mut wb = Workbook::new();
    let ws = wb.add_worksheet();

    // Aplicar anchos
    for &(col, ancho) in ANCHOS_COLUMNA {
        ws.set_column_width(col, ancho)?;
    }

    // Formato grupo (fila 0): centrado, wrap
    let fmt_grupo = Format::new()
        .set_align(FormatAlign::Center)
        .set_align(FormatAlign::VerticalCenter)
        .set_text_wrap();

    // Fila 0: grupos combinados
    ws.set_row_height(0, 18.0)?;
    ws.merge_range(0, 0, 0, 3, "Compareciente 1 / Vendedor", &fmt_grupo)?;
    ws.merge_range(0, 4, 0, 7, "Compareciente 2 / Comprador", &fmt_grupo)?;
    ws.merge_range(0, 8, 0, 19, "Vehículo", &fmt_grupo)?;
    ws.write_with_format(0, 20, "Datos OT", &fmt_grupo)?;

    // Formato encabezado (fila 1): borde thin top+bottom, centrado
    let fmt_enc = Format::new()
        .set_align(FormatAlign::Center)
        .set_align(FormatAlign::VerticalCenter)
        .set_border_top(FormatBorder::Thin)
        .set_border_bottom(FormatBorder::Thin);

    // Fila 1: encabezados
    ws.set_row_height(1, 30.0)?;
    for (col, &enc) in ENCABEZADOS.iter().enumerate() {
        ws.write_with_format(1, col as u16, enc, &fmt_enc)?;
    }

    // Formato datos: Calibri 11, izquierda
    let fmt_datos = Format::new()
        .set_font_name("Calibri")
        .set_font_size(11.0)
        .set_align(FormatAlign::Left);

    // Filas de datos (empiezan en fila 2)
    for (i, fila) in filas.iter().enumerate() {
        let row = (i + 2) as u32;
        let cols: Vec<String> = vec![
            fila.c1_rut.clone(),
            fila.c1_apellido_paterno.clone(),
            fila.c1_apellido_materno.clone().unwrap_or_default(),
            fila.c1_nombres.clone().unwrap_or_default(),
            fila.c2_rut.clone(),
            fila.c2_apellido_paterno.clone(),
            fila.c2_apellido_materno.clone().unwrap_or_default(),
            fila.c2_nombres.clone().unwrap_or_default(),
            fila.tasacion_fiscal.clone().unwrap_or_default(),
            fila.precio_venta.clone().unwrap_or_default(),
            fila.patente.clone().unwrap_or_default(),
            fila.tipo_vehiculo.clone().unwrap_or_default(),
            fila.marca.clone().unwrap_or_default(),
            fila.modelo.clone().unwrap_or_default(),
            fila.anio.clone().unwrap_or_default(),
            fila.color.clone().unwrap_or_default(),
            fila.motor.clone().unwrap_or_default(),
            fila.chasis.clone().unwrap_or_default(),
            fila.serie.clone().unwrap_or_default(),
            fila.vin.clone().unwrap_or_default(),
            fila.codigo_operacion.clone().unwrap_or_default(),
        ];

        for (col, valor) in cols.iter().enumerate() {
            ws.write_with_format(row, col as u16, valor.as_str(), &fmt_datos)?;
        }
    }

    // Crear directorio padre si no existe
    let path = std::path::Path::new(ruta);
    if let Some(padre) = path.parent() {
        if !padre.as_os_str().is_empty() {
            std::fs::create_dir_all(padre)?;
        }
    }

    wb.save(ruta)?;
    Ok(())
}

// ─── Nombre de planilla ───────────────────────────────────────────────────────

pub fn nombre_planilla(prefijo_sp: &str) -> String {
    if !prefijo_sp.is_empty() {
        format!("Repertorios Masivo SP {}.xlsx", prefijo_sp)
    } else {
        format!(
            "planilla_masivos_{}.xlsx",
            Local::now().format("%Y-%m-%d_%H-%M")
        )
    }
}
